agent/verify_job.go

```diff
@@ -0,0 +1,95 @@
+package agent
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/buildkite/agent/v3/internal/pipeline"
+)
+
+var ErrNoSignature = errors.New("job had no signature to verify")
+
+type invalidSignatureError struct {
+	underlying error
+}
+
+func newInvalidSignatureError(err error) *invalidSignatureError {
+	return &invalidSignatureError{underlying: err}
+}
+
+func (e *invalidSignatureError) Error() string {
+	return fmt.Sprintf("invalid signature: %v", e.underlying)
+}
+
+func (e *invalidSignatureError) Unwrap() error {
+	return e.underlying
+}
+
+func (e *invalidSignatureError) As(target any) bool {
+	_, ok := target.(*invalidSignatureError)
+	return ok
+}
+
+func (r *JobRunner) verificationFailureLogs(err error, behavior string) {
+	label := "WARNING"
+	if behavior == VerificationBehaviourBlock {
+		label = "ERROR"
+	}
+
+	r.logger.Warn("Job verification failed: %s", err.Error())
+	r.logStreamer.Process([]byte(fmt.Sprintf("⚠️ %s: Job verification failed: %s\n", label, err.Error())))
+
+	if behavior == VerificationBehaviourWarn {
+		r.logger.Warn("Job will be run without verification - this is not recommended. You can change this behavior with the `job-verification-no-signature-behavior` agent configuration option.")
+		r.logStreamer.Process([]byte(fmt.Sprintf("⚠️ %s: Job will be run without verification\n", label)))
+	}
+}
+
+func (r *JobRunner) verifyJob(verifier pipeline.Verifier) error {
+	step := r.conf.Job.Step
+
+	if verifier == nil {
+		if step.Signature != nil {
+			return fmt.Errorf("job %q was signed with signature %q, but no verification key was provided, so the job can't be verified", r.conf.Job.ID, step.Signature.Value)
+		}
+
+		return nil // no signature, no verifier, no problem
+	}
+
+	if _, ok := step.RemainingFields["matrix"]; ok {
```

Comment:
Since we need a semantic understanding of matrix vs non-matrix, perhaps `CommandStep` should grow a `Matrix` field (of type `any` I guess)?

Reply:
ooh, good call